hypothesis-python/src/hypothesis/internal/conjecture/data.py

```diff
@@ -923,11 +923,9 @@ def draw_boolean(self, p: float = 0.5, *, forced: Optional[bool] = None) -> bool
             # to write a byte to the data stream anyway so that these don't cause
             # difficulties when shrinking.
             if p <= 0:
-                self._cd.draw_bits(1, forced=0)
-                result = False
+                result = self._cd.draw_bits(1, forced=0 if forced is None else forced)
             elif p >= 1:
-                self._cd.draw_bits(1, forced=1)
-                result = True
+                result = self._cd.draw_bits(1, forced=1 if forced is None else forced)
```

Comment:
we didn't respect `forced` when p was "certain". oops!

Reply:
I think it should be an error, actually. I was too hasty in "fixing" this. This combination is simply not representable, similar to `draw_integer(0, 1, forced=2)`.

Another related interaction is that `p < 2^-64` is treated as `p = 0`. Forcing `True` here *should* be valid, but we'll have to modify the `if bits > 64:` conditional to e.g.` if bits > 64 and forced is None:` to consider this.

Reply:
well...making this an error turns out to be questionable in practice, because the table built in `Sampler` (sometimes?) includes an alternate_chance of 0.0, and we want to force this choice sometimes.

This means we very likely have an additional latent bug somewhere, because — as we discovered in this pull — forcing `True` with p=0 returned `False` instead, so this choice was not working as intended.

I'm looking into whether this can be fixed at the Sampler level, as this smells a bit like a bug in building the table (0 probability choices should be dropped). But maybe this indicates that getting this correct is more trouble than it's worth, and we should allow forcing p = 0, 1 choices in the opposite direction?